### PR TITLE
 fix: cap _get_large_negative at float32 range to avoid -inf under debug_infs with float64 inputs

### LINKED ISSUE #37422: dot_product_attention triggers FloatingPointError in convert_element_type with float64 mask and debug_infs=True
### Description

`jax.nn.dot_product_attention` triggers a `FloatingPointError` when `jax.debug_infs(True)` is enabled.

The inputs are all zeros with `float64` dtype, and the mask is a simple boolean mask. The expected attention output should be finite, but JAX reports an invalid `inf` value in `convert_element_type`.

Minimal reproducible example:

```
import sys

try:
    import jax
    import jax.numpy as jnp
    import numpy as np
except ImportError as e:
    print(f"missing dep: {e}")
    sys.exit(2)

def main() -> int:
    T, S, K, N, H = 5, 7, 3, 6, 11
    mask = np.r_[:S] <= S // 2

    try:
        jax.config.update("jax_enable_x64", True)
        with jax.debug_infs(True):
            out = jax.nn.dot_product_attention(
                np.zeros((T, N, H), dtype=np.float64),
                np.zeros((S, K, H), dtype=np.float64),
                np.zeros((S, K, H), dtype=np.float64),
                mask=mask,
            )
            out.block_until_ready()
        print("NOT REPRODUCED: no FloatingPointError with debug_infs=True")
        return 1
    except FloatingPointError as e:
        print(f"BUG REPRODUCED: FloatingPointError triggered: {e}")
        return 0
    except Exception as e:
        print(f"NOT REPRODUCED: different exception {type(e).__name__}: {e}")
        return 1

if __name__ == "__main__":
    raise SystemExit(main())
```

### System info (python version, jaxlib version, accelerator, etc.)

```
WARNING:2026-05-07 00:33:38,915:jax._src.xla_bridge:794: An NVIDIA GPU may be present on this machine, but a CUDA-enabled jaxlib is not installed. Falling back to cpu.
jax:    0.6.2
jaxlib: 0.6.2
numpy:  2.2.6
python: 3.10.20 (main, Mar 11 2026, 17:46:40) [GCC 14.3.0]
device info: cpu-1, 1 local devices
process_count: 1
platform: uname_result(system='Linux', node='MSI', release='5.15.167.4-microsoft-standard-WSL2', version='#1 SMP Tue Nov 5 00:21:55 UTC 2024', machine='x86_64')

nvidia-smi:
NVIDIA-SMI 545.23.06
Driver Version: 545.92
CUDA Version: 12.3
GPU: NVIDIA GeForce RTX 3080, 8192 MiB

Note: Reproduced on CPU fallback because CUDA-enabled jaxlib is not installed.
```

### PR DESCRIPTION
**Pull request by ShresthSamyak**
Fixes #37422

  ## Root cause

  When inputs are `float64`, logits are promoted to `float64` and
  `_get_large_negative(float64)` returns ~`-1.26e308`. This is a valid
  float64 value but overflows to `-inf` when cast to `float32` for softmax
  (`padded_logits.astype(np.float32)` in `_dot_product_attention_core`).
  `jax.debug_infs(True)` fires on that intermediate `-inf` in the
  `convert_element_type` op, before softmax can normalize it away.

  ## Fix

  Cap `_get_large_negative` at `float32` max (~`3.4e38`). The resulting
  value `~-2.38e38` is still large enough that `exp(x) ≈ 0` after softmax,
  so masked positions correctly receive zero probability. It has no effect
  on float32/float16/bfloat16 inputs since their `dtype_max` is already
  within float32 range.

  ## Test

  Added `testDotProductAttentionFloat64MaskDebugInfs` which reproduces the
  exact shape/dtype from the issue and asserts no `FloatingPointError` is
  raised.

### CODE REVIEW

File: tests/nn_test.py [hunk 1]
```diff
@@ -335,6 +335,24 @@ def testDotProductAttentionMask(self, mask_mode):
     self.assertAllClose(dV_ref, dV_ans, rtol=.01, atol=.01)
     self.assertAllClose(dbias_ref, dbias_ans, rtol=.02, atol=.02)
 
+  def testDotProductAttentionFloat64MaskDebugInfs(self):
+    # Regression test for https://github.com/jax-ml/jax/issues/37422
+    prev_x64 = jax.config.jax_enable_x64
```
Comment (jakevdp) [inline tests/nn_test.py]: Use `with jax.enable_x64():` to do this state management automatically.

**Review by jakevdp — CHANGES_REQUESTED** [review]
Looks good, just one request related to the test.

File: tests/nn_test.py [hunk 2]
```diff
@@ -335,6 +335,20 @@ def testDotProductAttentionMask(self, mask_mode):
     self.assertAllClose(dV_ref, dV_ans, rtol=.01, atol=.01)
     self.assertAllClose(dbias_ref, dbias_ans, rtol=.02, atol=.02)
 
+  def testDotProductAttentionFloat64MaskDebugInfs(self):
+    # Regression test for https://github.com/jax-ml/jax/issues/37422
+    with jax.enable_x64():
+      T, S, K, N, H = 5, 7, 3, 6, 11
+      mask = jnp.arange(S) <= S // 2
+      with jax.debug_infs(True):
+        out = nn.dot_product_attention(
```
Comment (jakevdp) [inline tests/nn_test.py]: To be explicit here, please use `with self.assertNoWarnings()` so we're not relying on pytest-only global warning configurations.

**Review by jakevdp — CHANGES_REQUESTED** [review]

**Review by jakevdp — COMMENTED** [review]
Looks great, thanks! Last thing: please squash all changes into a single commit.

**Review by jakevdp — APPROVED** [review]
